src/plugins/camera/camera_impl.cpp

```diff
@@ -932,15 +936,62 @@ void CameraImpl::process_camera_image_captured(const mavlink_message_t& message)
         _captured_request_cv.notify_all();
 
         std::lock_guard<std::mutex> lock(_capture_info.mutex);
-        if (_capture_info.last_advertised_image_index != -1 &&
-            _capture_info.last_advertised_image_index < capture_info.index &&
-            _capture_info.callback) {
-            _capture_info.last_advertised_image_index = capture_info.index;
-            const auto temp_callback = _capture_info.callback;
-            _parent->call_user_callback(
-                [temp_callback, capture_info]() { temp_callback(capture_info); });
+        if (_capture_info.last_advertised_image_index != -1) {
+            // Notify user if a new image has been captured.
+            if (_capture_info.last_advertised_image_index < capture_info.index) {
+                if (_capture_info.callback) {
+                    const auto temp_callback = _capture_info.callback;
+                    _parent->call_user_callback(
+                        [temp_callback, capture_info]() { temp_callback(capture_info); });
+                }
+
+                // Save captured indices that have been dropped
+                for (int i = _capture_info.last_advertised_image_index + 1; i < capture_info.index;
+                     ++i) {
+                    if (_capture_info.missing_image_retries.find(i) ==
+                        _capture_info.missing_image_retries.end()) {
+                        _capture_info.missing_image_retries[i] = 0;
+                    }
+                }
+
+                _capture_info.last_advertised_image_index = capture_info.index;
+            }
+
+            else if (auto it = _capture_info.missing_image_retries.find(capture_info.index);
+                     it != _capture_info.missing_image_retries.end()) {
+                if (_capture_info.callback) {
+                    const auto temp_callback = _capture_info.callback;
+                    _parent->call_user_callback(
+                        [temp_callback, capture_info]() { temp_callback(capture_info); });
+                }
+                _capture_info.missing_image_retries.erase(it);
+            }
+        }
+    }
+}
+
+void CameraImpl::request_missing_capture_info()
+{
+    std::lock_guard<std::mutex> lock(_capture_info.mutex);
+
+    for (auto it = _capture_info.missing_image_retries.begin();
```

Comment:
I think you want a `while` here :sweat_smile:

Reply:
There is this exact pattern a couple of times in this codebase with `for` but a `while` would also work, you're right, and might make it clearer.